### PR TITLE
FE-243: Auto-detect if non-priv mode should be used

### PR DESCRIPTION
**Pull request by pankaj-mandal**
When the jupyter_pachyderm plugin is used in an Determined AI root-less/daemon-less HPC container runtime (Singularity/Enroot/Apptatiner/Podman), automatically assume NONPRIV_CONTAINER=1 to enable /pfs to work.
This is accomplished by checking the environment variable DET_RESOURCES_TYPE for the value "slurm-job" which is set by Determined when launching such containers on an HPC cluster.

The following tests were done and found to be successful

setup
Built a docker image locally on my laptop which includes changes in this PR
Tagged it and pushed it to docker hub
Success criteria: Connect to pachyderm cluster (cluster 1) in Houston lab
Testing
Tested the docker image by running it locally. This was done without setting nonpriv option. Test was successful
Build Singularity and Podman images on casablanca and tested both of them without nonpriv option set. Test was successful.
Note:- The singularity and podman tests above was done as described in FE-80
Used the above built singularity image to run it in non-privileged mode

singularity run --no-mount=tmp --writable-tmpfs  --env NONPRIV_CONTAINER=1 notebooks-user-dev.sif
This test was successful

Tested by setting
export DET_RESOURCES_TYPE="slurm-job"
and running the command
singularity run --no-mount=tmp --writable-tmpfs notebooks-user-dev.sif
this test was successful too.
singularity-nonpriv-autodetect-mode

### CODE REVIEW

**Comment by pankaj-mandal** [comment]
Created this branch as the other one was having issues due to my email settings.

File: jupyter-extension/jupyterlab_pachyderm/env.py [hunk 1]
```diff
@@ -9,7 +9,8 @@
 SIDECAR_MODE = strtobool(os.environ.get("SIDECAR_MODE", "False").lower())
 MOUNT_SERVER_LOG_DIR = os.environ.get("MOUNT_SERVER_LOG_DIR") # defaults to stdout/stderr
 NONPRIV_CONTAINER = os.environ.get("NONPRIV_CONTAINER") # Unset assume --privileged container
```
Comment (jerryharrow) [inline jupyter-extension/jupyterlab_pachyderm/env.py:11]: Looks like this should perhaps be more like `PACHYDERM_EXT_DEBUG` so it is really a bool and not just a string.

`strtobool(os.environ.get("NONPRIV_CONTAINER", "False").lower())`

File: jupyter-extension/jupyterlab_pachyderm/mount_server_client.py [hunk 2]
```diff
@@ -37,8 +46,11 @@ def __init__(
             self.sock_path = ""
         self.client = AsyncHTTPClient()
         # non-prived container flag (set via -e NONPRIV_CONTAINER=1)
-        # TODO: Would be preferable to auto-detect this, but unclear how
+        # or use DET_RESOURCES_TYPE environment variable to auto-detect this.
         self.nopriv = NONPRIV_CONTAINER
+        self.determined_resources_type = DET_RESOURCES_TYPE
```
Comment (jerryharrow) [inline jupyter-extension/jupyterlab_pachyderm/mount_server_client.py:51]: What is the need for making an additional variable that is never referenced?

Why not just:

`self.nopriv = NONPRIV_CONTAINER or DET_RESOURCES_TYPE == SLURM_JOB`

**Comment by jerryharrow** [comment]
Looks good, but can you list the test cases you've verified?

**Comment by pankaj-mandal** [comment]
> Looks good, but can you list the test cases you've verified?

I will do some manual verification.

**Comment by pankaj-mandal** [comment]
> > Looks good, but can you list the test cases you've verified?
> 
> I will do some manual verification.

The following tests were done and found to be successful

**setup**
Built a docker image locally on my laptop which includes changes in this PR
Tagged it and pushed it to docker hub
Success criteria: Connect to pachyderm cluster (cluster 1) in Houston lab
Testing
Tested the docker image by running it locally. This was done without setting nonpriv option. Test was successful
Build Singularity and Podman images on casablanca and tested both of them without nonpriv option set. Test was successful.
Note:- The singularity and podman tests above was done as described in FE-80
Used the above built singularity image to run it in non-privileged mode

`singularity run --no-mount=tmp --writable-tmpfs  --env NONPRIV_CONTAINER=1 notebooks-user-dev.sif`
This test was successful

Tested by setting 
`export DET_RESOURCES_TYPE="slurm-job"`
and running the command
 `singularity run --no-mount=tmp --writable-tmpfs notebooks-user-dev.sif`
this test was successful too.
singularity-nonpriv-autodetect-mode

**Comment by pankaj-mandal** [comment]
There is one issue I noticed unrelated to this.
If there are multiple mount-server processes running, the mount does not work. Connection to pachyderm seems to go through but the mount does not work. Specifically the following line in mount_server_client.py fails
`mount_server_pid = int(mount_server_pid)`
this is because it is trying to cast string containing numbers to integer, but the string also contains some non numeric values to separate multiple pids. We can address this in separate JIRA.

**Comment by jerryharrow** [comment]
> If there are multiple mount-server processes running, the mount does not work. Connection to pachyderm seems to go through but the mount does not work. Specifically the following line in mount_server_client.py fails

Yes, but of course that isn't supposed to happen.   The mount_server_client.py checks for an existing mount-server and attempts to kill it before starting a new one.   Before #9389 (where I added -s to pgrep) with Singularity/Enroot it could observe mount-servers started by other users/sessions and that caused the failure.   With that fix, I believe it should be robust except when multiple mount-servers are started in the same container (which the code looks like the code attempts to prevent).    If you reproduce it, are we apparently are getting 2 mount-server processes inside the same session, so we need to diagnose how that happens to prevent the problem.

Another somewhat hackey fix would be only return only one mount-server pid, but we would have to figure out which is the right one.  With a fixed port it would be the first one started that would have allocated the port.  With  a unix domain socket, would then most-recent instance be the one serving the current /tmp/pfs.sock ?

**Comment by jerryharrow** [comment]
Looks like the build infrastructure is broken, you probably need to rebase.

**Comment by pankaj-mandal** [comment]
> Looks like the build infrastructure is broken, you probably need to rebase.

As of now it is upto date. @bbonenfant any idea what to do for the circleci?

**Comment by pankaj-mandal** [comment]
> > If there are multiple mount-server processes running, the mount does not work. Connection to pachyderm seems to go through but the mount does not work. Specifically the following line in mount_server_client.py fails
> 
> Yes, but of course that isn't supposed to happen. The mount_server_client.py checks for an existing mount-server and attempts to kill it before starting a new one. Before #9389 (where I added -s to pgrep) with Singularity/Enroot it could observe mount-servers started by other users/sessions and that caused the failure. With that fix, I believe it should be robust except when multiple mount-servers are started in the same container (which the code looks like the code attempts to prevent). If you reproduce it, are we apparently are getting 2 mount-server processes inside the same session, so we need to diagnose how that happens to prevent the problem.
> 
> Another somewhat hackey fix would be only return only one mount-server pid, but we would have to figure out which is the right one. With a fixed port it would be the first one started that would have allocated the port. With a unix domain socket, would then most-recent instance be the one serving the current /tmp/pfs.sock ?

I think in my case it is happening due to improperly removed containers, i.e. even after doing ctrlz and then terminating the container. It looks like the mount server processes are left behind. I have this situation for enroot for sure. But enroot is not working for me so far. So I guess we can keep it in the back burner and revisit once I have enroot working. Also we can in general check for mount-server processes before starting a new container and kill them if any processes are found.

File: jupyter-extension/jupyterlab_pachyderm/mount_server_client.py [hunk 3]
```diff
@@ -37,8 +46,10 @@ def __init__(
             self.sock_path = ""
         self.client = AsyncHTTPClient()
         # non-prived container flag (set via -e NONPRIV_CONTAINER=1)
-        # TODO: Would be preferable to auto-detect this, but unclear how
+        # or use DET_RESOURCES_TYPE environment variable to auto-detect this.
         self.nopriv = NONPRIV_CONTAINER
+        if DET_RESOURCES_TYPE == SLURM_JOB:
+            self.nopriv = 1
```
Comment (bbonenfant) [inline jupyter-extension/jupyterlab_pachyderm/mount_server_client.py:52]: I think adding a logging statement here that NONPRIV is being inferred would be helpful.

**Review by bbonenfant — APPROVED** [review]
Looks good

File: jupyter-extension/jupyterlab_pachyderm/mount_server_client.py (hunk 3 above)
Comment (pankaj-mandal) [reply jupyter-extension/jupyterlab_pachyderm/mount_server_client.py:52]: > I think adding a logging statement here that NONPRIV is being inferred would be helpful.

ok

**Pull request merged**